service-workers/service-worker/xhr-content-length.window.js

```diff
@@ -0,0 +1,71 @@
+// META: script=resources/test-helpers.sub.js
+
+let frame;
+
+promise_test(async (t) => {
+  const scope = "resources/empty.html";
+  const script = "resources/xhr-content-length-worker.js";
+  const registration = await service_worker_unregister_and_register(t, script, scope);
+  await wait_for_state(t, registration.installing, "activated");
+  frame = await with_iframe(scope);
+}, "Setup");
+
+promise_test(async t => {
+  const xhr = new frame.contentWindow.XMLHttpRequest();
+  xhr.open("GET", "test?type=no-content-length");
+  xhr.send();
+  await new Promise(resolve => {
+    xhr.onload = t.step_func(e => {
+      assert_equals(xhr.getResponseHeader("content-length"), null);
+      assert_false(e.lengthComputable);
+      assert_equals(e.total, 0);
+      assert_equals(e.loaded, xhr.responseText.length);
+      resolve();
+    });
+  });
+}, `Synthetic response without Content-Length header`);
+
+promise_test(async t => {
+  const xhr = new frame.contentWindow.XMLHttpRequest();
+  xhr.open("GET", "test?type=larger-content-length");
+  xhr.send();
+  await new Promise(resolve => {
+    xhr.onload = t.step_func(e => {
+      assert_equals(xhr.getResponseHeader("content-length"), "10000");
+      assert_true(e.lengthComputable);
+      assert_equals(e.total, 10000);
+      assert_equals(e.loaded, xhr.responseText.length);
+      resolve();
+    });
+  });
+}, `Synthetic response with Content-Length header with value larger than response body length`);
+
+promise_test(async t => {
+  const xhr = new frame.contentWindow.XMLHttpRequest();
+  xhr.open("GET", "test?type=double-content-length");
+  xhr.send();
+  await new Promise(resolve => {
+    xhr.onload = t.step_func(e => {
+      assert_equals(xhr.getResponseHeader("content-length"), "10000, 10000");
+      assert_true(e.lengthComputable);
+      assert_equals(e.total, 10000);
+      assert_equals(e.loaded, xhr.responseText.length);
+      resolve();
+    });
+  });
+}, `Synthetic response with two Content-Length headers value larger than response body length`);
```

Comment:
This test captured the main weirdness. Somehow Chrome and Firefox (not sure how to run service worker tests in Safari yet) return "`10000`" despite the synthetic response having two `Content-Length` headers.

Reply:
@mfalken what do you think about this behavior? Chromium generates net::ERR_RESPONSE_HEADERS_MULTIPLE_CONTENT_LENGTH for multiple content-length headers in response coming from the network[1], but that seems missing in the service worker case[2].

1: https://source.chromium.org/chromium/chromium/src/+/master:net/http/http_stream_parser.cc;l=1043?q=RESPONSE_HEADERS_MULTIPLE_CONTENT_LENGTH%20file:net%2F&ss=chromium
2: https://source.chromium.org/chromium/chromium/src/+/master:content/common/service_worker/service_worker_loader_helpers.cc;l=89;drc=56c01dcef367cdbda4e2592a9d0d897f8272fdc7

Reply:
Without really knowing about the "response smuggling attack" cited in the comment, I would tend to say Chrome should treat responses from the SW and from the network as much as possible, so I think we should throw the same network error.

Reply:
1. I think I disagree, as we it's only about handing a Response over a thread boundary.
2. This does not create a network error in Chrome today for a response coming from the network. Multiple `Content-Length` headers with the same value are fine.
3. If we did want to error for multiple `Content-Length` headers in a synthetic response where the headers have different values, we have to do a lot more work in writing those kind of conditions down.
4. The problem here is that Chrome (and Firefox) return(s) `10000` rather than `10000, 10000`. In Firefox that happens because the header lookup code apparently special cases `Content-Length`...

Reply:
@yutakahirano @mfalken is there a way we can move this forward? I suppose I can add a test for the network equivalent (which will likely also fail, but not because the response is rejected, as per above), though it's a bit out-of-scope for what I was going for it does seem good to test. Anything else?

Reply:
So I went ahead and added a test for this and it seems that Chrome doesn't have any problems with duplicates from the network (but Firefox and Safari do). Convincing enough?

Reply:
I don't have a strong opinion here and would like to defer to @mflaken. I'm fine with the change overall.

Reply:
I filed https://github.com/w3c/ServiceWorker/issues/1571 as follow-up as this matches the current specification.